In [246]:
#Importing necessary libraries
import numpy as np
import random
import time

# Answer 1

## a(i) Use gradient descent to find minima for $f(x) = x^2 + 3x + 4$

 $$ g = \frac{\partial f(x)}{\partial x} = \frac{\partial (x^2 + 3x + 4)}{\partial x} = 2x + 3 $$

In [247]:
def gradient( x ):
    return 2*x + 3

In [248]:
def gradient_descent(gradient, init_, learn_rate, n_iter = 5000, tol=1e-06):
    x = init_
    for _ in range(n_iter):
        delta = -learn_rate * gradient(x)
        if np.all(abs(delta)<=tol):
            break
        x += delta
    return round(x*1000)/1000

In [249]:
minx = gradient_descent(gradient, 4 , 0.05 )
print("The point at which function attains minimum value is ",minx)

The point at which function attains minimum value is  -1.5


$f(x)=x^2+3x+4$ <br> 
$f'(x)={2}x+3$ <br>
$f'(x)=0 {\;}{\;} when{\;}{\;} x=-1.5$ <br>
$f''(x)=2 >0{\;}{\;} when{\;}{\;} x=-1.5$<br>
$\therefore$ It is postive semidefinite and hence it is the minima.

In [250]:
def f(x):
    return x*x + 3*x + 4
print("Minimum value of function is ",f(minx))

Minimum value of function is  1.75


## a(ii) Use gradient descent to find minima for $ x^4 - 3x^2 + 2x $

$$ g = \frac{\partial f(x)}{\partial x} = \frac{\partial (x^4 - 3x^2 + 2x)}{\partial x} = 4x^3 - 6x $$

In [251]:
def gradient( x ):
    return 4*(x**3) - 6*x + 2

In [252]:
def gradient_descent(gradient, init_, learn_rate, n_iter = 5000, tol=1e-06):
    x = init_
    for _ in range(n_iter):
        delta = -learn_rate * gradient(x)
        if np.all(abs(delta)<=tol):
            break
        x += delta
    return round(x*1000)/1000

In [253]:
minx = gradient_descent( gradient, 0 , 0.01 )
print("The point at which function attains minimum value is ",minx)

The point at which function attains minimum value is  -1.366


$f(x)=x^4-3x^2+2x $ <br> 
$f'(x)={4}x^3-{6}x+2$ <br>
$f'(x)=4x^3-6x+2=0 \Rightarrow  x = 1, -1/2\pm \sqrt3/2$<br>
$f''(1)=6 >0$  <br>
$f''(-1/2+\sqrt3/2) = -4.39 < 0 \Rightarrow Rejected{\;}as{\;}it{\;}cannot{\;}be{\;}minima$<br>
$f''(-1/2-\sqrt3/2) = 16.39 > 0 $<br>
$f(1)=0 \Rightarrow Local{\;}minima$<br>
$f(-1/2-\sqrt3/2)=f(-1.366)=-4.8480 \Rightarrow Global{\;}minima$<br>
$\therefore$ It is postive semidefinite and hence it is the minima.

In [254]:
def f( x ):
    return (x**4) - 3*(x**2) + 2*x
print("Minimum value of function is ",f(minx))

Minimum value of function is  -4.848076206064


# 1(b).  Gradient function to calculate gradients for a linear regression 

$$y = aX + b$$
$$L = \frac{\sum{(y - y\_predict)}^2}{N}$$<br>
$$\frac{\partial (L)}{\partial a} = -\frac{2}{N}*\sum(X(y - y\_predict))$$<br>
$$\frac{\partial (L)}{\partial b} = -\frac{2}{N}*\sum(y - y\_predict) $$

### Gradient : the direction of the steepest change in function value

In [255]:
def gradient( X, y, y_predict):
    #length of the dataset .ie. No. of instances
    n = X.shape[0] 
    gradient_b = -(2/n)*sum(y - y_predict) # As per the gradient formula derived
    gradient_a = -(2/n)*np.dot(X.T,(y - y_predict)) # As per the gradient formula derived
    return gradient_b, gradient_a

# 1(c). Use gradient descent to find the optimal parameters relating X with y

In [256]:
# Dataset generation
np.random.seed(0)
X = 2.5*np.random.randn(10000)+1.5 
res=1.5*np.random.randn(10000) 
y = 2+0.3*X + res

Gradient Descent Algorithm <br>
$\bf for{\;}k=0,1,2\dots (or{\;}until{\;}convergence)$ <br>
$\bf do $<br> 
${\;}{\;}{\;}{\;}{\;}{\;}g_k = \nabla f(x_k)$<br>
${\;}{\;}{\;}x_{k+1} = x_k - t_k g_k$ <br>
$\bf end{\;}for$<br>



In [257]:
def gradient_descent(X, y, learn_rate = 0.05, max_iter = 1000, tol = 1e-06):
    a = 0
    b = 0
    
    gradient_a = 0 #gradient of a
    gradient_b = 0 ##gradient of b
    
    
    for i in range( max_iter ):
        y_predict = a*X + b
        g = gradient(X, y , y_predict) #as defined in 1(b)
        
        gradient_b = g[0]
        gradient_a = g[1]
        
#         print('gradient_a: ', gradient_a, '    gradient_b: ', gradient_b)
        
        delta_a = -1*learn_rate*(gradient_a)
        delta_b = -1*learn_rate*(gradient_b)
        
        if (np.all(np.abs(delta_a) <= tol) & np.all(np.abs(delta_b) <= tol)):
            break
        a = a + delta_a
        b = b + delta_b
    return (round(a,2),round(b,2))

In [258]:
timegd = time.time()
a_calculated, b_calculated = gradient_descent(X, y, 0.005, 1000, 1e-06)
timegd = time.time() - timegd

In [259]:
print("Predicted value using gradient descent linear algorithm ")
print("Optimal parameters relating X and y are: a=",a_calculated," b=",b_calculated)

Predicted value using gradient descent linear algorithm 
Optimal parameters relating X and y are: a= 0.3  b= 2.02


## 1.(d) Minibatch Gradient Descent

Initialize $\mathbf{w}$ as $\mathbf{w^0}$<br>
Set a learning rate ${\mathbf{n}}_t$<br>
Choose a random number B uniformly randomnly, and select $\{i_1, i_2,...,i_B\}\in \{1,2,...N\}$<br> 
$\bf for{\;}i=0,1,2\dots (or{\;}until{\;}convergence)$ <br>
$    {\;}{\;}{\;}{\;}{\;}    {\mathbf{g}} =\sum_{b=1}^B \mathbf{g}_{i_b}$<br>
${\;}{\;}{\;}{\;}{\;}\mathbf{w}^{(t+1)} = \mathbf{w}^{(t)} -{\mathbf{n}_t}{\mathbf{g}}$

In [260]:
def minibatch_gradient_descent(X, y, batchsize, learn_rate = 0.01, max_iter = 5000, tol = 1e-06):
    a = 0
    b = 0
    
    gradient_a = 0
    gradient_b = 0
    
    n = X.shape[0]
    iterations = 0
    for i in range( max_iter ):
        iterations = iterations + 1

        B = np.random.randint(0,n-batchsize-1) #B: block size taken from [B] to [B+batchsize]
        X_temp = X[B :B + batchsize] #current batch sample
        y_temp = y[B :B + batchsize] #current batch output
        
        n_temp = X_temp.shape[0]  #size of current batch taken
        y_predict = a*X_temp + b  #predicted value of current batch
        
        g = gradient(X_temp, y_temp , y_predict) #as defined in 1(b)
        
        gradient_b = g[0]
        gradient_a = g[1]
        
        delta_a = -1*learn_rate*(gradient_a)
        delta_b = -1*learn_rate*(gradient_b)
        
        if (np.all(np.abs(delta_a) <= tol) & np.all(np.abs(delta_b) <= tol)):
            break
        a = a + delta_a
        b = b + delta_b
    return (round(a,2),round(b,2),iterations)

In [261]:
print("Minibatch Gradient Descent")

a_predicted, b_predicted, iterations = minibatch_gradient_descent(X, y, 1024)

print(a_predicted, b_predicted, iterations)

Minibatch Gradient Descent
0.3 2.01 5000


## Stochastic Gradient Descent

Initialize $\mathbf{w}$ as $\mathbf{w^0}$<br>
Set a learning rate ${\mathbf{n}}_t$<br>
$\bf for{\;}i=0,1,2\dots (or{\;}until{\;}convergence)$ <br>

$    {\;}{\;}{\;}{\;}{\;}$Choose a random number $ k\in \{1,2,...N\}$<br>
$    {\;}{\;}{\;}{\;}{\;}    {\mathbf{g}} =\mathbf{g}_{k}$<br>
${\;}{\;}{\;}{\;}{\;}\mathbf{w}^{(t+1)} = \mathbf{w}^{(t)} -{\mathbf{n}_t}{\mathbf{g}}$

In [262]:
def stochastic_gradient_descent(X, y, learn_rate = 0.05, max_iter = 10000, tol = 1e-06):
    a = 0
    b = 0
    
    gradient_a = 0
    gradient_b = 0
    batchsize = 1
    n = X.shape[0]
    iterations = 0
    for i in range( max_iter ):
        iterations = iterations + 1

        B = np.random.randint(0,n-1) #B: block size taken from [B] to [B+batchsize]
        X_temp = X[B :B + batchsize] #current batch sample
        y_temp = y[B :B + batchsize] #current batch output
        
        n_temp = X_temp.shape[0]  #size of current batch taken
        y_predict = a*X_temp + b  #predicted value of current batch
        
        g = gradient(X_temp, y_temp , y_predict) #as defined in 1(b)
        
        gradient_b = g[0]
        gradient_a = g[1]
        
        delta_a = -1*learn_rate*(gradient_a)
        delta_b = -1*learn_rate*(gradient_b)
        
        if (np.all(np.abs(delta_a) <= tol) & np.all(np.abs(delta_b) <= tol)):
            break
        a = a + delta_a
        b = b + delta_b
    return (round(a,2),round(b,2),iterations)


### Since stochastic gradient descent gives random outcomes. So I am trying to run the algorithm multiple times and find the optimal execution when it converges and note the time taken in such a case

In [263]:
for i in range(1000):
    timeSGD = time.time()
    a_predicted, b_predicted, iterations = stochastic_gradient_descent(X, y, 0.005, 1000, 1e-06)
    timeSGD = time.time() - timeSGD
    if ((a_predicted in [0.29, 0.3, 0.31]) and (b_predicted in [1.99, 2.0, 2.01])):
        break
print("Stochastic Gradient Descent Best Outcome")
print("a=",a_predicted,",b=", b_predicted,"iterations=", iterations,",time taken=",timeSGD)

Stochastic Gradient Descent Best Outcome
a= 0.3 ,b= 1.99 iterations= 1000 ,time taken= 0.09245705604553223


# 1.(e)  Time performance on our data

## Performance Estimation
#### Here I estimate the performance of MiniBatch Gradient Descent, Stochastic Gradient (SGD) and Gradient Descent (GD) and thereafter compare their runtimes.
#### I have taken batch sizes in power of 2, -ie- 1,2,4,8,16,32,64... because it is computationally faster.

In [264]:
for i in range(0,12):
    batchsize = (2**i)
    timeMGD = time.time()
    optimal_batchsize = batchsize
    a1,b1,itr1 = minibatch_gradient_descent(X,y,batchsize,0.005, 1000,1e-06) #(X, y, batchsize, learn_rate = 0.01, max_iter = 5000, tol = 1e-04)
    timeMGD = time.time() - timeMGD
    if ((a1 in [0.29, 0.3, 0.31]) and (b1 in [1.99, 2.0, 2.01])):
        break

In [265]:
print("Time taken by Minibatch Gradient Descent:",timeMGD)
print("Time taken by Batch Gradient Descent:",timegd)
print("Time taken by Stochastic Gradient Descent:",timeSGD)

Time taken by Minibatch Gradient Descent: 0.11353182792663574
Time taken by Batch Gradient Descent: 5.345445394515991
Time taken by Stochastic Gradient Descent: 0.09245705604553223


## Performance Interpretation
### (i) It is clear that Stochastic Gradient Descent (SGD) performs the best among all of them. It takes the minimum time for running on the given data set. This is so because it takes only a single data point for gradient update. The quantified proof is given below in the runtime comparison of SGD with all others.
#### $$Batch\; Gradient\; Descent > MiniBatch\; Gradient\; Descent > Stochastic\; Gradient\; Descent$$ 


In [266]:
print("Comparison between Minibatch and Batch GD:")
print("Time taken by Minibatch Gradient Descent:",timeMGD)
print("Time taken by Batch Gradient Descent:",timegd)
print("% reduction in time by Minibatch over Batch GD: ", (timegd-timeMGD)*100/timegd)

Comparison between Minibatch and Batch GD:
Time taken by Minibatch Gradient Descent: 0.11353182792663574
Time taken by Batch Gradient Descent: 5.345445394515991
% reduction in time by Minibatch over Batch GD:  97.87610162395241


In [267]:
print("Comparison between Batch Gradient Descent and SGD:")
print("Time taken by Batch Gradient Descent:",timegd)
print("Time taken by Stochastic Gradient Descent:",timeSGD)
print("% reduction in time by SGD over Batch GD: ", (timegd-timeSGD)*100/timegd)

Comparison between Batch Gradient Descent and SGD:
Time taken by Batch Gradient Descent: 5.345445394515991
Time taken by Stochastic Gradient Descent: 0.09245705604553223
% reduction in time by SGD over Batch GD:  98.27035823543561


In [268]:
print("Comparison between Minibatch Gradient Descent and SGD:")
print("Time taken by Minibatch Gradient Descent:",timeMGD)
print("Time taken by Stochastic Gradient Descent:",timeSGD)
print("% reduction in time by SGD over Minibatch Gradient Descent: ", (timeMGD-timeSGD)*100/timeMGD)
#Note: If the percentage turns out negative, please run the entire notebook again.

Comparison between Minibatch Gradient Descent and SGD:
Time taken by Minibatch Gradient Descent: 0.11353182792663574
Time taken by Stochastic Gradient Descent: 0.09245705604553223
% reduction in time by SGD over Minibatch Gradient Descent:  18.562875508991215


### (ii) Optimal batch size for Mini Batch Gradient Descent

In [269]:
print("Optimal batchsize = ",optimal_batchsize)

Optimal batchsize =  64


# Answer 2

$\mathbf{Let{\;}us{\;}define{\;}Events:}$<br>
$\bullet cold$ : person has cold<br>
$\bullet lung$ : person has lung disease<br>
$\bullet cough$ : person has cough<br>
$\bullet smoke$ : person smokes<br>

## 2(a) Calculate the probability that someone has both cold and a fever

Given data:  
$(i){\;}{\;} P(cold)=0.02$<br>
$(ii){\;}{\;} P(fever | cold)=0.307$<br><br>
Formula Used:<br>
Conditional probability formula  <br>
$P(A|B)=\frac{P(A\cap B)}{P(B)} $
where A, B are any 2 events.<br>
By cross multiplication, we get <br>
$P(A\cap B)= P(A|B)*P(B)$<br><br>
Solution:<br>
Substitute A = fever and B = cold in the above formula, we get <br>
$P(fever \cap cold) = P(cold)* P(fever|cold)$ <br>
$P(fever \cap cold) = 0.02 * 0.307 $<br>
$P(fever \cap cold) = 0.00614 $<br>
$ \therefore Answer = P(fever \cap cold) = 0.00614 $

## 2(b)  Calculate the probability that someone who has a cough has a cold.

$\mathbf{To{\;}find{\;}:}$<br>
$P(cough|cold) = ?$

$\mathbf{Formula{\;}used:}$<br>
$$ Bayes'{\;}{\;}Theorem{\;}P(cold|cough) =\frac{P(cold \cap cough)}{P(cough)}{\;}{\;}$$

$\mathbf{Solution :}$<br><br>
By theorem of Total Probability<br>
$P(lung{\;}) = P(lung\cap smoke) + P(lung\cap smoke^c)$<br> 
$P(lung{\;}) = {P(smoke)}{ P(lung\;|smoke)} + P{(smoke^c)}{P(lung\;|smoke)}$<br>
$P(lung{\;}) = 0.2 * 0.1009 + 0.8 * 0.001 $<br>
$P(lung{\;}) = 0.02018 + 0.0008 $<br>
$P(lung{\;}) =0.02098  \ldots (1) $

Probability of Intersection of 2 or more events <br>
$P(A\cap B \cap C) =P((A\cap B)\cap C) =P(C|(A\cap B)) P(A\cap B) = P(C|(A\cap B))({P(B|A)}{P(A)} ={P(A)}{P(B|A)}{P(C|A\cap B)} \ldots (2)$<br>
$P(cough)=P(cough\cap lung\cap cold) + P(cough\cap lung\cap cold^c)+ P(cough\cap lung^c\cap cold)+ P(cough\cap lung^c \cap cold^c) \ldots (3)$<br><br>
Evaluating the expression on RHS by using eqn (2) on each term of RHS <br>

$P(cough\cap lung\cap cold) = P(cough |(lung\cap cold) * P(lung|cold) * P(cold)$<br>
$P(cough\cap lung\cap cold) = 0.0525 * 0.02098 * 0.02$<br>
$P(cough\cap lung\cap cold) = 0.000315749 \ldots\ldots (4)$<br>


$P(cough\cap lung\cap cold^c) = P(cough |(lung\cap cold^c) * P(lung|cold^c) * P(cold^c)$<br>
$P(cough\cap lung\cap cold^c) = 0.505 * 0.02098 * 0.98$<br>
$P(cough\cap lung\cap cold^c) = 0.010383002 \ldots\ldots (5)$<br>

$P(cough\cap lung^c\cap cold) = P(cough |(lung^c\cap cold) * P(lung^c|cold) * P(cold)$<br>
$P(cough\cap lung^c\cap cold) = 0.505 * 0.97902 * 0.02$<br>
$P(cough\cap lung^c\cap cold) = 0.00988102 \ldots\ldots (6)$<br>


$P(cough\cap lung^c\cap cold^c) = P(cough |(lung^c\cap cold^c) * P(lung^c|cold^c) * P(cold)$<br>
$P(cough\cap lung^c\cap cold^c) = 0.01 * 0.97902 * 0.98$<br>
$P(cough\cap lung^c\cap cold^c) = 0.00988102 \ldots\ldots (7)$<br>


Substituting the values from equation (4),(5),(6),(7) into (3),<br>
$P(cough) = 0.000315749  + 0.010383002 + 0.00988102 + 0.00988102 $<br>
$P(cough) = 0.030181249 \ldots\ldots(8)$<br>


$$ P(cold|cough) =\frac{P(cold \cap cough)}{P(cough)}{\;}{\;}$$<br>
$$ P(cold|cough) = P(cold\cap cough\cap lung) + P(cold\cap cough\cap lung^c)$$<br>
$$ P(cold|cough) = P(cough |(lung\cap cold)*P(lung|cold)*P(cold) + P(cough|(lung^c\cap cold))*P(lung^c|cold)*P(cold)$$<br>
$$ P(cold|cough) = 0.0525 * 0.02098 * 0.02 + 0.505 * 0.97902 * 0.02 $$<br>
$$ P(cold|cough) = 0.000315749 + 0.00988102 $$<br>
$$ P(cold|cough) = 0.3380857764 $$<br>
$$ \therefore Answer{\;}= P(cold|cough) = 0.3380857764 $$<br>

# Answer 3

## Derive the MLE for the parameters of a k-sided multinomial distribution


Let 
<br>
$n = $ No. of times we repeat the experiement 
<br>
$k = $ No. of possible outcomes for each individual experiment


$Y$ :  Random variable denoting the outcome for each possible outcomes  
$Y : {Y_1 , Y_2, Y_3 \ldots Y_k}$ 
<br>
where 
<br>
$Y_1$ denotes first possible outcome<br>
$Y_2$ denotes second possible outcome<br>
$\vdots$<br>
$Y_k$ denotes kth possible outcome<br>

$$P(Y_1=x_1, Y_2=x_2,...,Y_k=x_k | {\theta_1^{x_1}}{\theta_2^{x_2}}\ldots{\theta_k^{x_k}})=\bigl({\frac{n!}{{x_1!}{x_2!}\ldots{x_k!}}}\bigr) {{\theta_1^{x_1}}{\theta_2^{x_2}}\ldots{\theta_k^{x_k}}}$$ 


where <br>
$x_1$ : No. of times first outcome occured <br>
$x_2$ : No. of times second outcome occured <br>
$x_3$ : No. of times third outcome occured <br>
$\vdots$<br>
$x_k$ : No. of times kth outcome occured <br>
$$\therefore x_1 + x_2 + x_3 + \cdots + x_k = n$$ <br>
$$\sum_{i=0}^n x_i = {n}$$


and <br>
$\theta_1$ : probability of having first outcome <br>
$\theta_2$ : probability of having second outcome <br>
$\theta_3$ : probability of having third outcome <br>
$\vdots$ <br>
$\theta_k$ : probability of having kth outcome <br>
$$\therefore \theta_1 + \theta_2 + \theta_3 + \cdots + \theta_k = 1$$ <br>
$$\sum_{i=1}^k \theta_i = {1}$$

Therefore the likelihood <br> 
$ L(\theta) $<br>
$= L(x_1, x_2, \ldots,x_n | \theta) $<br>
$= L(x_1, x_2, \ldots,x_n | {\theta_1}{\theta_2}\ldots{\theta_k}) $<br>

$=\bigl({\frac{n!}{{x_1!}{x_2!}\ldots{x_k!}}}\bigr) {{\theta_1^{x_1}}{\theta_2^{x_2}}\ldots{\theta_k^{x_k}}}$

Let us define a new function <br> $ LL(\theta) = log {\;}L(\theta) $<br>
Substituting the value from above,<br>
$ LL(\theta) = \log {\;} \Bigl(\bigl({\frac{n!}{{x_1!}{x_2!}\ldots{x_k!}}}\bigr) {{\theta_1^{x_1}}{\theta_2^{x_2}}\ldots{\theta_k^{x_k}}}\Bigr) $ <br>
Expanding logarithm on each term <br>
$ LL(\theta) = \log {\;} ( n! \theta_1^{x_1} \theta_2^{x_2} \ldots \theta_k^{x_k})  - \log (x_1! x_2! \ldots x_k!)$ <br>
$ LL(\theta) = \log{\;}(n!) -  \sum_{i=1}^k x_i \log\theta_i   - \log \sum_{i=1}^k (x_i!)$ <br>


Adding a regularization parameter $\alpha$, to avoid $\theta$ taking too large values <br>
$\alpha$: Langrangian parameter for regularization <br>
Now, the equation becomes $ Regularized{\;} Log{\;} Likelihood{\;} \theta$ abbreviated as $RLL(\theta,\alpha)$<br>
$ RLL(\theta,\alpha) = LL(\theta) + \alpha ( 1 - \sum_{i=1}^k \theta_i )$




To find the maximum of a function, we differentiate it with respect to $\theta$ and equating it to zero <br>
$$ \frac{\partial RLL(\theta,\alpha) } {\partial \theta} = \Bigl( \frac{\partial RLL(\theta,\alpha) } {\partial \theta_1}, \frac{\partial RLL(\theta,\alpha)}{\partial \theta_2},\ldots\frac{\partial RLL(\theta,\alpha)}{\partial \theta_k}\Bigr)$$



$$\frac{\partial RLL(\theta,\alpha) } {\partial \theta_1} =\frac{\partial \log{\;}(n!) } {\partial \theta_1} + \frac{\partial\sum_{i=1}^k x_i \log\theta_i}{\partial \theta_1} - \frac{\partial\log \sum_{i=1}^k (x_i!)}{\partial \theta_1} + \frac{\partial(\alpha ( 1 - \sum_{i=1}^k \theta_i ))}{\partial \theta_1} $$<br>


On solving the above equation, we get <br>
$$\frac{\partial RLL(\theta,\alpha)}{\partial \theta_1} =\frac{x_1}{\theta_1} - \alpha = 0 $$
$Similarly$<br>
$$\frac{\partial RLL(\theta,\alpha)}{\partial \theta_2} =\frac{x_2}{\theta_2} - \alpha = 0 $$
$$\vdots$$
$$\frac{\partial RLL(\theta,\alpha)}{\partial \theta_k} =\frac{x_k}{\theta_k} - \alpha = 0 $$

$$\therefore\theta_1 = \frac{x_1}{\alpha},$$<br> 
$Similarly$<br>
$$ \theta_2 = \frac{x_2}{\alpha},$$<br>
$$ \vdots $$<br>
$$ \theta_k = \frac{x_k}{\alpha}  $$

$$ \therefore \sum_{i=1}^k \theta_i = 1 \Rightarrow \alpha = n $$

$$ \frac{\partial^2 RLL(\theta,\alpha)}{\partial \theta^2} = 
\begin{bmatrix}
\frac{\partial^2 RLL(\theta,\alpha)}{\partial \theta_1^2} & \frac{\partial^2 RLL(\theta,\alpha)}{\partial\theta_1\partial\theta_2 } & \ldots & \frac{\partial^2 RLL(\theta,\alpha)}{\partial\theta_1\partial\theta_k} \\
\frac{\partial^2 RLL(\theta,\alpha)}{\partial \theta_2\partial\theta_1} & \frac{\partial^2 RLL(\theta,\alpha)}{\partial\theta_2^2 } & \ldots & \frac{\partial^2 RLL(\theta,\alpha)}{\partial\theta_2\partial\theta_k}\\
\vdots & & \ddots \\
\frac{\partial^2 RLL(\theta,\alpha)}{\partial \theta_k\partial\theta_1} & \frac{\partial^2 RLL(\theta,\alpha)}{\partial \theta_k\partial\theta_2} & \ldots & \frac{\partial^2 RLL(\theta,\alpha)}{\partial\theta_k^2}\\\\
\end{bmatrix}
$$

$$ \frac{\partial^2 RLL(\theta,\alpha)}{\partial \theta^2} = 
\begin{bmatrix}
\frac{\partial (\frac{x_1}{\theta_1} - \alpha)}{\partial \theta_1} & \frac{\partial(\frac{x_2}{\theta_2} - \alpha)}{\partial\theta_1} & \ldots & \frac{\partial(\frac{x_k}{\theta_k} - \alpha)}{\partial\theta_1} \\
\frac{\partial(\frac{x_1}{\theta_1} - \alpha)}{\partial \theta_2} & \frac{\partial(\frac{x_2}{\theta_2} - \alpha)}{\partial\theta_2 } & \ldots & \frac{\partial(\frac{x_k}{\theta_k} - \alpha)}{\partial\theta_2}\\
\vdots & & \ddots \\
\frac{\partial(\frac{x_1}{\theta_1} - \alpha)}{\partial \theta_k} & \frac{\partial(\frac{x_2}{\theta_2} - \alpha)}{\partial \theta_k} & \ldots & \frac{\partial(\frac{x_k}{\theta_k} - \alpha)}{\partial\theta_k}\\\\
\end{bmatrix}
$$

$$ \frac{\partial^2 RLL(\theta,\alpha)}{\partial \theta^2} = 
\begin{bmatrix}
\frac{-x_1}{\theta_1^2} & 0 & \ldots & 0 \\
0 & \frac{-x_2}{\theta_2^2} & \ldots & 0 \\
\vdots & & \ddots \\
0 & 0 & \ldots & \frac{-x_k}{\theta_k^2} \\\\
\end{bmatrix}
$$

$$ \frac{-x_i}{\theta_i^2} \lt 0 {\;} $$
$$\forall{\;}1\leq i\leq k$$ 

Since the eigen values of the above matrix are negative. Hence, it is Negative Semi-Definite Matrix.<br>
Therefore, the likelihood function is going to give us a Maximum.<br>


$$Thus,{\;}the{\;}Maximum{\;}Likelihood{\;}Estimate{\;}(MLE){\;}of{\;}k-fold{\;}multinomial{\;}distribution{\;}is{\;}$$
$$\therefore\theta_1 = \frac{x_1}{n}, \theta_2 = \frac{x_2}{n}, \ldots \theta_k = \frac{x_k}{n}  $$

In [270]:
print("End of assignment")

End of assignment
